### PR TITLE
[Feature Request]: Batch deprecated object cleanup #978

### PR DESCRIPTION
**Pull request by rjzhb**
### What problem does this PR solve?

This PR aims to enhance the cleanup mechanism by addressing two main drawbacks identified in issue #978:

Improving the mechanism for recording the oldest visible timestamp to avoid unnecessary access to txn_map.
Optimizing the cleanup process to allow bulk deletion of files and records, potentially saving significant file IO operations. Additionally, bulk deletion from the buffer manager can reduce lock conflicts.

Issue link:#978

### Type of change

-New Feature (non-breaking change which adds functionality)

### Basic Thinking
Use set (red black tree data structure) to save oldestTs, replacing the original txn_map
Change the deletion of one by one to batch one-time deletion. After completing this, it is necessary to modify the unit testing of buffer_obj

### CODE REVIEW

File: src/storage/buffer/buffer_obj.cpp [hunk 1]
```diff
@@ -197,10 +198,12 @@ void BufferObj::SetAndTryCleanup() {
...
             LOG_TRACE(fmt::format("Remove file: {}", file_name));
-            file_worker_->CleanupFile();
-            LOG_TRACE(fmt::format("Remove from buffer: {}", file_name));
-            buffer_mgr_->RemoveBufferObj(file_name);
-            LOG_TRACE(fmt::format("Removed file and buffer: {}", file_name));
+
+            String file_path = fmt::format("{}/{}", *(file_worker_->file_dir_), *(file_worker_->file_name_));
+            buffer_mgr_->AddFilePath(file_path);
+            LOG_TRACE(fmt::format("Add the file to be deleted from buffer: {}", file_name));
+            buffer_mgr_->AddBufferObjPath(file_name);
```
Comment (JinHai-CN) [inline src/storage/buffer/buffer_obj.cpp:205]: AddBufferObjectForDeletion

File: src/storage/buffer/file_worker/file_worker.cpp [hunk 2]
```diff
@@ -119,4 +119,18 @@ void FileWorker::CleanupFile() {
...
+void FileWorker::CleanupFilesInBulk(const Vector<String> &file_paths) {
+    LocalFileSystem fs;
+    for (unsigned long i = 0; i < file_paths.size(); i++) {
+        const auto &path = file_paths[i];
+        if (fs.Exists(path)) {
+            LOG_TRACE(fmt::format("Cleaning up file: {}", path));
+            fs.DeleteFile(path);
+            LOG_TRACE(fmt::format("Cleaned file: {}", path));
+        } else {
+            LOG_TRACE(fmt::format("Cleanup: File {} not found for deletion", path));
```
Comment (JinHai-CN) [inline src/storage/buffer/file_worker/file_worker.cpp:131]: Move the comment on why the path doesn't exist here.

File: src/storage/buffer/file_worker/file_worker.cpp [hunk 3]
```diff
@@ -119,4 +119,18 @@ void FileWorker::CleanupFile() {
     LOG_TRACE(fmt::format("Cleaned file: {}", file_path));
 }
 
+void FileWorker::CleanupFilesInBulk(const Vector<String> &file_paths) {
+    LocalFileSystem fs;
+    for (unsigned long i = 0; i < file_paths.size(); i++) {
```
Comment (JinHai-CN) [inline src/storage/buffer/file_worker/file_worker.cpp:124]: Why not use for-range?

File: src/storage/buffer/file_worker/file_worker.cppm [hunk 4]
```diff
@@ -59,6 +59,7 @@ public:
 
     void CloseFile();
 
+    static void CleanupFilesInBulk(const Vector<String> &file_paths);
```
Comment (JinHai-CN) [inline src/storage/buffer/file_worker/file_worker.cppm:62]: BulkCleanup

File: src/storage/buffer/buffer_manager.cpp [hunk 5]
```diff
@@ -83,6 +83,26 @@ void BufferManager::RemoveBufferObj(const String &file_path) {
     LOG_TRACE(fmt::format("Erased buffer object: {}", file_path));
 }
 
+void BufferManager::RemoveBufferObjsInBulk(const Vector<String> &paths_to_delete) {
+    std::unique_lock<std::mutex> lock(w_locker_);
+    deprecated_array_.clear();
```
Comment (JinHai-CN) [inline src/storage/buffer/buffer_manager.cpp:88]: No need guarded by locker.

File: src/storage/buffer/buffer_manager.cpp [hunk 6]
```diff
@@ -83,6 +83,26 @@ void BufferManager::RemoveBufferObj(const String &file_path) {
...
+
+    for (unsigned long i = 0; i < paths_to_delete.size(); i++) {
```
Comment (JinHai-CN) [inline src/storage/buffer/buffer_manager.cpp:90]: Why not use for-range?

**Comment by JinHai-CN** [comment]
As for the improvement of 'oldest visible timestamp'.

We only need to record the begin timestamp and txn id of each committed txn. And check if the txn id is in wait_flush_txns.

File: src/storage/buffer/buffer_manager.cpp [hunk 7]
```diff
@@ -83,6 +83,26 @@ void BufferManager::RemoveBufferObj(const String &file_path) {
...
+        const auto &file_path = paths_to_delete[i];
+        LOG_TRACE(fmt::format("Erasing buffer object: {}", file_path));
+        deprecated_array_.emplace_back(buffer_map_[file_path]);
```
[the remark this answers is not in the document]
Comment (JinHai-CN) [reply src/storage/buffer/buffer_manager.cpp:94]: I mean emplace_back data to deprecated_array_ don't need locker guard. But buffer_map_ need it.
You may erase file from buffer_map_ in another loop.

File: src/storage/buffer/file_worker/file_worker.cpp [hunk 8]
```diff
@@ -13,6 +13,7 @@
 // limitations under the License.
 
 module;
+#include "base64.hpp"
```
Comment (JinHai-CN) [inline src/storage/buffer/file_worker/file_worker.cpp:16]: Do we really need this header?

File: src/storage/buffer/buffer_manager.cpp [hunk 9]
```diff
@@ -13,14 +13,14 @@
 // limitations under the License.
 
 module;
+#include "base64.hpp"
```
Comment (JinHai-CN) [inline src/storage/buffer/buffer_manager.cpp:16]: Do we really need this header here?

File: src/storage/buffer/file_worker/file_worker.cpp (hunk 8 above)
Comment (rjzhb) [reply src/storage/buffer/file_worker/file_worker.cpp:16]: I think so

Comment (rjzhb) [reply src/storage/buffer/file_worker/file_worker.cpp:16]: like src/storage/fast_rough_filter/probabilistic_data_filter.cpp and src/storage/fast_rough_filter/min_max_data_filter.cpp

**Pull request merged**